packages/astro/src/webcontainer-files/filesmap.ts

```diff
@@ -0,0 +1,218 @@
+import type { AstroIntegrationLogger } from 'astro';
+import glob from 'fast-glob';
+import fs from 'node:fs';
+import path from 'node:path';
+import { z } from 'zod';
+import type { Files } from '../types';
+import { EXTEND_CONFIG_FILEPATH, IGNORED_FILES } from './constants';
+
+const configSchema = z
+  .object({
+    extends: z.string(),
+  })
+  .strict();
+
+export class FilesMapGraph {
+  constructor(private _nodes = new Map<string, FilesMap>()) {}
+
+  getFilesMapByFolder(folder: string) {
+    try {
+      const resolvedPath = fs.realpathSync(folder);
+
+      return this._nodes.get(resolvedPath);
+    } catch {
+      return undefined;
+    }
+  }
+
+  allFilesMap() {
+    return this._nodes.values();
+  }
+
+  updateFilesMapByFolder(folder: string, logger: AstroIntegrationLogger) {
+    const resolvedPath = fs.realpathSync(folder);
+
+    let node = this._nodes.get(resolvedPath);
+
+    if (!node) {
+      node = new FilesMap(resolvedPath);
+      this._nodes.set(resolvedPath, node);
+    }
+
+    node.update(this._nodes, logger);
+  }
+}
+
+export class FilesMap {
```

Comment:
I created this class to make it easier to understand the graph structure and encapsulate the logic around:
* Cycle detection (which should not throw or cause an infinite loop)
* The conversion from a node in that graph to a `Files` entity that can be JSON serialized later